wnfs-common/src/blockstore/carblockstore.rs

```diff
@@ -0,0 +1,360 @@
+            factory.current_size + block_size as usize + block_size_bytes.len() > max_size
+        }
+        // If there is no max size the data is never too big
+        else {
+            false
+        };
+
+        // If there is no CAR or we don't have enough space left to fit this data
+        if factory.current_car.is_none() || data_too_big {
+            // Rotate the CAR to make room
+            factory.rotate()?;
+        }
+
+        // Grab a mutable reference to the CarFile's BufWriter
+        let writable_car: &mut BufWriter<File> = factory.current_car.as_mut().unwrap();
+
+        // Write the block size to the current CAR file
+        writable_car.write_all(&block_size_bytes)?;
+        // Write the CID of the block
+        writable_car.write_all(&cid_bytes)?;
+        // Write the contents of the block
+        writable_car.write_all(&bytes)?;
+        // Flush the Writer to ensure that those bytes are actually written
+        writable_car.flush().unwrap();
+
+        // Denote LocationInCar
+        let loc = LocationInCar {
+            car_number: factory.car_number,
+            offset: factory.current_size,
+        };
+
+        // Increment the size of the current CAR
+        factory.current_size += block_size as usize + block_size_bytes.len();
+
+        // Grab write lock and insert the <Cid, LocationInCar> pairing into the HashMap
+        self.index
+            .write()
+            .map_err(|e| anyhow!("{e}: couldn't get write lock"))?
+            .insert(cid, loc);
+
+        // TODO (organizedgrime)
+        // might there be scenarios we want to do this given that we don't know when someone might serialize?
+        // factory.finish()?;
+
+        // Return generated CID for future retrieval
+        Ok(cid)
+    }
+}
+
+#[derive(Serialize, Deserialize)]
+struct CarHeader {
+    version: i128,
+    roots: Vec<Cid>,
+}
+
+impl CarHeader {
+    pub fn new() -> Self {
+        Self {
+            version: 1,
+            roots: Vec::new(),
+        }
+    }
+
+    pub fn to_bytes(&self) -> Vec<u8> {
+        let roots = Ipld::List(self.roots.clone().into_iter().map(Ipld::Link).collect());
+        let header_ipld: Ipld = ipld!({
+          "version": self.version,
+          "roots": roots,
+        });
+        DagCborCodec.encode(&header_ipld).unwrap()
+    }
+
+    pub fn from_bytes(bytes: &[u8]) -> Self {
+        let header_ipld: Ipld = DagCborCodec.decode(bytes).unwrap();
+        let Ipld::Integer(version) = header_ipld.get("version").unwrap() else { panic!() };
+        let Ipld::List(roots) = header_ipld.get("roots").unwrap() else { panic!() };
+
+        let roots: Vec<Cid> = roots
+            .iter()
+            .map(|ipld| {
+                let Ipld::Link(cid) = ipld else { panic!() };
+                *cid
+            })
+            .collect();
+
+        CarHeader {
+            version: *version,
+            roots,
+        }
+    }
```

Comment:
There is a `dagcbor::{encode, decode}` functions in `encoding.rs` that help with this.